Improve accessibility on story block with keyboard shortcuts

Tug opened the pull request:
This PR adds keyboard support for the story block and more generally improves accessibility on the block.
This is a follow up of https://github.com/Automattic/jetpack/pull/16852


#### Changes proposed in this Pull Request:

This is an important refactor of the Story Block as we're moving away from managing the player's state within the player component itself to using a global (redux) store. This allows the story player to be unmounted and remounted somewhere else on the page.

This should be fixing the z-index issues as we now have a true modal component (appended at the end of the DOM with a high z-index).

When displaying a story in a modal we're focusing on the story player itself as well as trapping the keyboard in order to prevent tabbing on other elements of the page. On exit we refocus on the inlined payer element.

While playing controls can be iterated using tab/shit+tab and img alt and title can be read by screen readers. This should offer a slightly better overall experience for screen reader users.

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
- Test the changes on a .org and wpcom site
- Make sure there is no regression regarding story loading and playback
- Try playing a story using your keyboard: tab to focus on player, space to play/pause, escape to leave fullscreen
- Try from an android device
- Try from an iphone
- Try using voiceover on your mac and iphone
- Try using talkback on android

aforcier commented on projects/plugins/jetpack/changelog/update-story-block-keyboard-shortcuts:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: compat
+
+Story Block: Improve accessibility and resolve z-index issues when playing in fullscreen.
```
We've been leaving the Story block out of the release notes since it's not available on the web yet and untestable. You can update the changelog such that the comment won't be added to the release notes by using the Comment field, like I did here: https://github.com/Automattic/jetpack/pull/19176/files

Also we'd want to update the PR description not to have release notes.

aforcier commented:
@Tug I haven't tested everything yet but I ran into an issue I wanted to mention early. On my test site, applying this patch and loading the blog roll (which has multiple story posts) causes a weird scrolling issue - sometimes it auto-scrolls down to some story, sometimes down and back up to the top of the page.

It happens every time for me, on desktop and on Android Chrome. Here's a gif - I'm not interacting with the page in any way except for the initial refresh:

![stories-scrolling-issue](https://user-images.githubusercontent.com/9613966/115331535-10619700-a1d1-11eb-9130-37088a3422ed.gif)

I'll leave the patch applied on my sandbox so you can test the behavior on my sandbox site.

aforcier commented on projects/plugins/jetpack/extensions/blocks/story/player/lib/wait-media-ready.js:
```diff
@@ -22,18 +23,14 @@ export default async function waitMediaReady( mediaElement, fullLoad = false ) {
 				requestHeaders.append( 'Content-Type', mediaElement.type );
 			}
-			return fetch( videoRequest, {
+			const response = await fetch( videoRequest, {
```
I'm back to seeing endless loading animations on video-first stories a lot (https://github.com/Automattic/story-block/issues/12), I guess it might be because of this change?

aforcier replied:
Nevermind, I'm not able to reproduce this anymore, and per our conversation this shouldn't have affected the poster image loading.

aforcier commented:
Confirming that [the scrolling issue I reported earlier](https://github.com/Automattic/jetpack/pull/17342#issuecomment-822939689) is now fixed 👍 

I've tested the patch on a self-hosted site and on WP.com, on Android, iOS, Mac Chrome and Edge. The keyboard shortcuts work well on desktop, and everything seems to be working normally on mobile so far 🎉 

@Tug do I remember you saying that this patch should also resolve a few of our theme compatibility issues? Should this be tested on some of those themes as well?

Also, I'll spend some time testing with voiceover/talkback in the next few days, and in the meantime defer to @enejb for an actual code review 😀

enejb commented:
I tested this here https://ephemeral-enej-20210422.atomicsites.blog/2021/04/22/rock/ 

I think overall this works really well. Nice work! 
I was able to use the `enter` into a storey and exit one. As well as pause and start it and use left and right arrows to navigate the stories. 

A few things that I notice that I think could be fixed in the following PRs.
1. Running the page above using https://web.dev/measure/ to see if there are any accessibility issues. 
It came across https://web.dev/aria-hidden-focus/ error. I think the fix there is to add tabindex="-1" to controls. that are hidden. Although I didn't have an issue with my controls going missing when I tested it on the latest FF. 

I think more can be done to improve accessibility. 
2. I wasn't able to `tab` into the other controls that are present on the overlay. (Next and previous buttons) We should show a focus indicator as well. Right now this part seems a bit hidden. 
3. Maybe we can use `aria-group` to tell the reader more, how many stories are there? I wasn't able to get a voiceover to Read the image Alt titles. I think https://wp.stories.google/stories/tips-web-stories-editor/ does a good job there. 
4. It would be really great if the image alt titles were generated using text that the user entered over the image.

enejb commented on projects/plugins/jetpack/extensions/blocks/story/player/progress-bar.js:
```diff
@@ -1,22 +1,40 @@
-import { range } from 'lodash';
+import { useSelect } from '@wordpress/data';
+
+export const ProgressBullet = ( { key, playerId, index, disabled, isSelected, onClick } ) => {
```
❤️

jeherve reviewed (changes requested):
You seem to have a minor linting error, and multiple files that are now error-free so they can be removed from the exclusion list! 🥳

Tug commented:
Thank you all for the reviews! I'll try to address all the feedback I got :)

> @Tug do I remember you saying that this patch should also resolve a few of our theme compatibility issues? Should this be tested on some of those themes as well?

Yes this should be fixing all of the z-index issues. I did test on TT and it indeed fixed it. Feel free to test on other themes

> Running the page above using https://web.dev/measure/ to see if there are any accessibility issues...

Thanks for pointing out those red flags on this service (1.), I'll look into it. Those might be false positives though as I really tried to tie my accessibility attribute to how VoiceOver was behaving. In general I agree that accessibility can be further improved and I'm willing to try different things.

Regarding tabbing (2.), it should be available if you have VoiceOver on, I don't think it works without it (or maybe buttons are simply not highlighted I need to check).

(3.) Not sure what you mean here? Apply the `group` role to some parents of all stories? Stories can be placed anywhere on the page so that doesn't seem feasible.. Maybe you meant group on the slides? I think that doesn't work as voiceover should be describing what's visible only and other slides are not visible. We do have a `tablist` role on the progressbar that list all the slides we can access though. 

VoiceOver should be reading `alt` and `title` (if you tab on it), maybe I have a regression due to my lint fixes, will try again 👍 

(4.) Yep, that should be working as the mobile apps set the image `alt` using the text added to the image.


> You seem to have a minor linting error, and multiple files that are now error-free so they can be removed from the exclusion list! 🥳

👀

jeherve commented:
I'm moving this out of the milestone since it seems that it's still in progress, and since we're approaching code freeze.

Tug commented:
@enejb I fixed the `:focus-visible` styles not showing up on some controls, so tabbing should be better when VoiceOver is off now.

I looked into the issue reported by `https://web.dev/measure/`, it's actually not related to the story block it seems, there's a `aria-hidden=true` attribute on the `<div id="page">` for some reason and I think that might be the issue.

Also tested voiceover again on a story that has images with alt and title attributes and could have them read by the screenreader (one need to tab into the figure so it works)

> Should we be adding tests for the new Redux store, Actions etc?

I definitely could, will see if I can add a few without it being a time sink. I added a bit of documentation in the meantime so one can get a little bit more familiar with the codebase.

Finally, I simplified some of the logic regarding video loading and added a way to detect when the video is buffering. In the end we shouldn't try to fetch a video using `fetch`, we should let the browser handle the fetching and caching as it's supposed to be. It can make the video playback choppy on slow connections, but I guess that should rather be mitigated with better video compression on the server (and maybe even [adaptative streaming](https://developer.mozilla.org/en-US/docs/Web/Media/DASH_Adaptive_Streaming_for_HTML_5_Video) if we want to go further)

sdixon194 commented:
Did some testing on this and had some issues depending on the browser:

**Firefox**

- There was no focus indicator when tabbing onto the story
- When focusing on the play/pause button, using spacebar had some odd behavior (this wasn't present on Chrome). Gif is too big, [but you can see it here](https://d.pr/i/HULgyH).

***Mobile***

Firefox on mobile (on an Android device) was a pretty rough experience. I had lot of difficulty opening and then closing the story, but I'm not sure that's related to this PR. 

**Safari**

- Wasn't able to focus on each element - I can cycle through the close button, the video itself, but not the track bar or the play/pause button - it jumps back up to the address bar. Not sure if that's expected.

**Chrome**

Seemed fine! Was able to tab through and see the indicator/select each element. No weird issue with spacebar.

I'm not sure how much of this is related to longer term browser issues or this PR specifically, but wanted to bring them up and get your thoughts!

Tug commented:
@sdixon194 Thanks for testing. Indeed there was an obscure issue on firefox where any video element would trap the focus when tabbing, I've fixed that now.

Regarding difficulties in tabbing on firefox and safari, are you using macos? Could you check that "Use keyboard navigation to move focus between controls" is checked on your [keyboard settings](https://support.apple.com/en-us/HT204434#fullkeyboard)?

I'll try the story block on FF for android, thanks for pointing it out.

sdixon194 commented:
>Regarding difficulties in tabbing on firefox and safari, are you using macos? Could you check that "Use keyboard navigation to move focus between controls" is checked on your keyboard settings?

Ah thanks, that did it for me and I can tab through and focus properly on both now! 

>Thanks for testing. Indeed there was an obscure issue on firefox where any video element would trap the focus when tabbing, I've fixed that now.

Hmm still having some trouble with this one. In Firefox, when play/pause is focused on, using spacebar kind of stutters the playback instead of playing/pausing fully. [This is the post](https://sdixon.blog/2021/04/28/test-taking-photo/) I was testing with.

Tug commented:
> **when play/pause is focused on**, using spacebar kind of stutters the playback instead of playing/pausing fully.

Ohhh right, that was the part I somehow missed when I read your feedback, makes total sense since we bind space on the whole dialog for playing and pausing, and space is also the default for pressing a button. Fixing this now! THanks

Tug commented:
Should be good now :)

enejb commented:
I spend some more time testing this on Latest FF, Safari and Chrome 
And it worked as expected for me! Nice work!

Tug commented:
Deployed on WPCOM with changeset ID 225203
